text/0000-fallible-allocation.md

```diff
@@ -0,0 +1,237 @@
+
+* [Linux](https://lore.kernel.org/lkml/CAHk-=wh_sNLoz84AUUzuqXEsYH35u=8HV3vK-jbRbJ_B-JjGrg@mail.gmail.com/) and other OS Kernel environments.
+* Mid-range embedded environments (e.g. CPU firmware, bootloaders, etc.)
+* High-reliability processes (e.g. init)
+* Services which allocate in response to untrusted user input
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+When a developer is writing code in an environment where `abort()`ing (or calling a dedicated OOM handler) is not an appropriate strategy for dealing with allocation failure scenarios, they should consider disabling the `infallible_allocation` feature.
+If the developer is writing in a hosted environment, they will disable the feature on the `std` crate.
+If the developer is writing in a bare-metal environment, they will instead disable the feature on the `alloc` crate.
+This should be presented as an exceptional situation - the vast majority of developers will want the default here.
+
+This will remove a large number of functions - many standard library functions implicitly allocate and do not handle OOM scenarios.
+
+Note also that this will either require the developer to download an additional sysroot via `rustup` or perform a local build, depending on the developer's target.
+
+## Kernel Example
+For example, those writing a kernel likely do not want to `abort()` just because memory allocation failed.
+Therefore, it would be preferable to them to have the compiler reject
+
+```
+fn do_thing(x: usize) {
+  other_fn(Box::new(x))
+}
+```
+
+and would prefer they write
+
+```
+fn do_thing(x: usize) -> Result<(), AllocErr> {
+  Box::try_new(x).map(other_fn)
+}
+```
+
+(or more likely, project the `AllocErr` into a local error type).
+
+To prevent the first example from compiling, the developer would add to `Cargo.toml`'s dependencies section:
+
+```
+alloc = { default-features = false }
+```
+
+If the developer is using other default features of `alloc`, they may need to add those explicitly.
+After this, the first example will fail to compile, but the second will succeed.
+
+## Hosted Example
+If the developer is authoring a high-reliability process that absolutely cannot crash (e.g. `init`), they will want the hosted equivalent.
+For the same code examples in the Kernel section above, they would instead add
+
+```
+std = { default-features = false }
+```
+
+Adding any additional features they wish to depend on explicitly.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Moving infallible allocation behind a feature
+All functions in `alloc` directly depending on the global OOM handler will be gated on the crate feature `infallible_allocation`.
+Any functions directly calling allocation APIs in either `alloc` or `std` will either propagate that error to its caller in a handleable way or be moved behind the crate feature `infallible_allocation`.
+
+## Suppressing annotations in `rustdoc`
```

Comment:
cc @rust-lang/rustdoc 

I think that this is a fair ask for an unstable internal-use-only feature, but just tagging y'all in.

Reply:
This does not need a new rustdoc flag. `doc(cfg(...))` is added manually wherever it exists, the functions would just not add that attribute.

https://github.com/rust-lang/rust/pull/79341 would change that but it also has a separate mechanism for opting out. It doesn't need to be address in this RFC.

Reply:
Oh, right. Even better, thanks!

Reply:
Thanks for the catch. I've updated the RFC to no longer suggest this.